File: src/expr.hs
-- Integer Expression AST/Evaluator: Haskell --

{--
  define an algebraic type representing the abstract syntax tree
  for our simple expression language:
  an Expr is either a Const (containing an integer value),
  or a Plus (of two Exprs),
  or a Sub (of two Exprs),
  or a Mul (of two Exprs),
  or a Div (of two Exprs)
--}
data Expr = Const Int
          | Plus Expr Expr
          | Sub Expr Expr
          | Mul Expr Expr
          | Div Expr Expr

{--
  to evaluate an Expr, we pattern-match against its top-level structure
  and produce the associated value of a Const, or evaluate recursively otherwise
--}
eval :: Expr -> Int
eval (Const n) = n
eval (Plus e1 e2) = (eval e1) + (eval e2)
eval (Sub e1 e2) = (eval e1) - (eval e2)
eval (Mul e1 e2) = (eval e1) * (eval e2)
eval (Div e1 e2) = (eval e1) `div` (eval e2)

-- evaluate a simple expression represented as a literal Expr
main :: IO ()
main = let e = Mul (Const 3) (Plus (Const 5) (Const 2)) in
    putStrLn $ show $ eval e

File: src/expr.rs
// Integer Expression AST/Evaluator: Rust


/* define an algebraic type representing the abstract syntax tree
 * for our simple expression language:
 * an Expr is either a Const (containing an integer value),
 * or a Plus (of two Exprs),
 * or a Sub (of two Exprs),
 * or a Mul (of two Exprs),
 * or a Div (of two Exprs)
 *
 * we need to Box the Exprs in the recursive types (that is, put
 *   them behind a pointer): otherwise we'd have an infinite-sized structure!
 */
pub enum Expr {
    Const(i32),
    Plus(Box<Expr>, Box<Expr>),
    Sub(Box<Expr>, Box<Expr>),
    Mul(Box<Expr>, Box<Expr>),
    Div(Box<Expr>, Box<Expr>),
}

impl Expr {
    // pattern-match the top-level and recurse to evaluate
    pub fn eval(&self) -> i32 {
        match *self {
            Expr::Const(n) => n,
            Expr::Plus(ref e1, ref e2) => e1.eval() + e2.eval(),
            Expr::Sub(ref e1, ref e2) => e1.eval() + e2.eval(),
            Expr::Mul(ref e1, ref e2) => e1.eval() * e2.eval(),
            Expr::Div(ref e1, ref e2) => e1.eval() / e2.eval(),
        }
    }
}

fn main() {
    let e = Expr::Mul(
        Box::new(Expr::Const(3)),
        Box::new(
            Expr::Plus(Box::new(Expr::Const(5)), Box::new(Expr::Const(2))))
    );

    println!("{}", e.eval());
}
